File: currency_converter.py
#!/usr/bin/env python3

import sys
import json
import os
from datetime import datetime
from typing import Optional, Dict
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
import requests
from decimal import Decimal

console = Console()

CURRENCY_CODES = {
    'USD': 'US Dollar', 'EUR': 'Euro', 'GBP': 'British Pound', 'JPY': 'Japanese Yen',
    'AUD': 'Australian Dollar', 'CAD': 'Canadian Dollar', 'CHF': 'Swiss Franc',
    'CNY': 'Chinese Yuan', 'NZD': 'New Zealand Dollar', 'SEK': 'Swedish Krona',
    'KRW': 'South Korean Won', 'SGD': 'Singapore Dollar', 'NOK': 'Norwegian Krone',
    'MXN': 'Mexican Peso', 'INR': 'Indian Rupee', 'RUB': 'Russian Ruble',
    'ZAR': 'South African Rand', 'TRY': 'Turkish Lira', 'BRL': 'Brazilian Real',
    'TWD': 'Taiwan Dollar', 'DKK': 'Danish Krone', 'PLN': 'Polish Zloty',
    'THB': 'Thai Baht', 'IDR': 'Indonesian Rupiah', 'HUF': 'Hungarian Forint',
    'CZK': 'Czech Koruna', 'ILS': 'Israeli Shekel', 'CLP': 'Chilean Peso',
    'PHP': 'Philippine Peso', 'AED': 'UAE Dirham', 'COP': 'Colombian Peso',
    'SAR': 'Saudi Riyal', 'MYR': 'Malaysian Ringgit', 'RON': 'Romanian Leu',
    'DOP': 'Dominican Peso', 'BGN': 'Bulgarian Lev', 'HKD': 'Hong Kong Dollar',
    'HRK': 'Croatian Kuna', 'PKR': 'Pakistani Rupee', 'EGP': 'Egyptian Pound',
    'ISK': 'Icelandic Króna', 'VND': 'Vietnamese Dong', 'NGN': 'Nigerian Naira'
}

API_BASE_URL = 'https://api.exchangerate-api.com/v4/latest'

class CurrencyConverter:
    def __init__(self):
        self.history_file = "conversion_history.json"
        self.rates_cache_file = "rates_cache.json"

    def load_cached_rates(self) -> dict:
        """Load cached exchange rates from file."""
        try:
            with open(self.rates_cache_file, 'r') as f:
                return json.load(f)
        except FileNotFoundError:
            return {}

    def save_cached_rates(self, rates: dict):
        """Save exchange rates to cache file."""
        with open(self.rates_cache_file, 'w') as f:
            json.dump(rates, f)

    def save_to_history(self, amount: float, from_currency: str, to_currency: str, result: float):
        """Save conversion to history file."""
        try:
            with open(self.history_file, 'r') as f:
                history = json.load(f)
        except FileNotFoundError:
            history = []

        history.append({
            'timestamp': datetime.now().isoformat(),
            'from_amount': amount,
            'from_currency': from_currency,
            'to_currency': to_currency,
            'result': result
        })

        with open(self.history_file, 'w') as f:
            json.dump(history, f, indent=2)

    def show_history(self):
        """Display conversion history in a table."""
        try:
            with open(self.history_file, 'r') as f:
                history = json.load(f)
        except FileNotFoundError:
            console.print("[yellow]No conversion history found.[/yellow]")
            return

        table = Table(title="Conversion History")
        table.add_column("Date", style="cyan")
        table.add_column("From", style="green")
        table.add_column("To", style="green")
        table.add_column("Result", style="yellow")

        for entry in history:
            table.add_row(
                datetime.fromisoformat(entry['timestamp']).strftime('%Y-%m-%d %H:%M:%S'),
                f"{entry['from_amount']} {entry['from_currency']}",
                entry['to_currency'],
                f"{entry['result']:.2f}"
            )

        console.print(table)

    def list_currencies(self):
        """Display available currencies in a table."""
        table = Table(title="Available Currencies")
        table.add_column("Code", style="cyan")
        table.add_column("Currency Name", style="green")

        for code, name in sorted(CURRENCY_CODES.items()):
            table.add_row(code, name)

        console.print(table)

    def convert(self, amount: float, from_currency: str, to_currency: str, offline: bool = False) -> float:
        """Convert currency using real-time or cached rates."""
        try:
            if offline:
                rates = self.load_cached_rates()
                if not rates:
                    console.print("[red]No cached rates available. Please run in online mode first.[/red]")
                    sys.exit(1)
                rate = rates.get(f"{from_currency}_{to_currency}")
                if not rate:
                    console.print(f"[red]No cached rate found for {from_currency} to {to_currency}[/red]")
                    sys.exit(1)
                return amount * rate
            else:
                response = requests.get(f"{API_BASE_URL}/{from_currency}")
                if response.status_code != 200:
                    console.print("[red]Error: Unable to fetch exchange rates.[/red]")
                    sys.exit(1)

                data = response.json()
                if 'rates' not in data:
                    console.print("[red]Error: Invalid response from exchange rate API.[/red]")
                    sys.exit(1)

                rate = data['rates'].get(to_currency)
                if rate is None:
                    console.print(f"[red]Error: Rate not available for {from_currency} to {to_currency}[/red]")
                    sys.exit(1)

                # Cache the rate
                rates = self.load_cached_rates()
                rates[f"{from_currency}_{to_currency}"] = rate
                self.save_cached_rates(rates)

                return amount * rate
        except requests.exceptions.RequestException as e:
            console.print(f"[red]Error: Unable to connect to exchange rate service. {str(e)}[/red]")
            sys.exit(1)
        except Exception as e:
            console.print(f"[red]Error: {str(e)}[/red]")
            sys.exit(1)

def display_result(amount: float, from_currency: str, to_currency: str, result: float):
    """Display conversion result in a beautiful format."""
    table = Table(show_header=False, box=None)
    table.add_row(
        f"[green]{amount:,.2f}[/green]",
        f"[yellow]{from_currency}[/yellow]",
        "→",
        f"[green]{result:,.2f}[/green]",
        f"[yellow]{to_currency}[/yellow]"
    )
    
    panel = Panel(
        table,
        title="Currency Conversion Result",
        border_style="blue"
    )
    console.print(panel)

@click.group()
def cli():
    """Modern Currency Converter CLI Tool"""
    pass

@cli.command()
@click.option('--amount', type=float, required=True, help='Amount to convert')
@click.option('--from', 'from_currency', required=True, help='Source currency code')
@click.option('--to', 'to_currency', required=True, help='Target currency code')
@click.option('--save', is_flag=True, help='Save conversion to history')
@click.option('--offline', is_flag=True, help='Use cached exchange rates')
def convert(amount: float, from_currency: str, to_currency: str, save: bool, offline: bool):
    """Convert currency from one to another"""
    converter = CurrencyConverter()
    
    try:
        # Convert currencies to uppercase
        from_currency = from_currency.upper()
        to_currency = to_currency.upper()
        
        # Check if currencies are supported
        if from_currency not in CURRENCY_CODES:
            console.print(f"[red]Error: Currency {from_currency} is not supported[/red]")
            sys.exit(1)
        if to_currency not in CURRENCY_CODES:
            console.print(f"[red]Error: Currency {to_currency} is not supported[/red]")
            sys.exit(1)
        
        # Perform conversion
        result = converter.convert(amount, from_currency, to_currency, offline)
        
        # Display result
        display_result(amount, from_currency, to_currency, result)
        
        # Save to history if requested
        if save:
            converter.save_to_history(amount, from_currency, to_currency, result)
            console.print("[green]Conversion saved to history.[/green]")
            
    except Exception as e:
        console.print(f"[red]Error: {str(e)}[/red]")
        sys.exit(1)

@cli.command()
def list():
    """List all available currencies"""
    converter = CurrencyConverter()
    converter.list_currencies()

@cli.command()
def history():
    """Show conversion history"""
    converter = CurrencyConverter()
    converter.show_history()

if __name__ == '__main__':
    cli() 

File: currency_converter.js
#!/usr/bin/env node

const { program } = require('commander');
const axios = require('axios');
const chalk = require('chalk');
const ora = require('ora');
const { table } = require('table');
const fs = require('fs').promises;
const path = require('path');
require('dotenv').config();

const CACHE_FILE = 'rates_cache.json';
const HISTORY_FILE = 'conversion_history.json';
const API_BASE_URL = 'https://api.exchangerate-api.com/v4/latest';

const CURRENCY_CODES = {
    USD: 'US Dollar', EUR: 'Euro', GBP: 'British Pound', JPY: 'Japanese Yen',
    AUD: 'Australian Dollar', CAD: 'Canadian Dollar', CHF: 'Swiss Franc',
    CNY: 'Chinese Yuan', NZD: 'New Zealand Dollar', SEK: 'Swedish Krona',
    KRW: 'South Korean Won', SGD: 'Singapore Dollar', NOK: 'Norwegian Krone',
    MXN: 'Mexican Peso', INR: 'Indian Rupee', RUB: 'Russian Ruble',
    ZAR: 'South African Rand', TRY: 'Turkish Lira', BRL: 'Brazilian Real',
    TWD: 'Taiwan Dollar', DKK: 'Danish Krone', PLN: 'Polish Zloty',
    THB: 'Thai Baht', IDR: 'Indonesian Rupiah', HUF: 'Hungarian Forint',
    CZK: 'Czech Koruna', ILS: 'Israeli Shekel', CLP: 'Chilean Peso',
    PHP: 'Philippine Peso', AED: 'UAE Dirham', COP: 'Colombian Peso',
    SAR: 'Saudi Riyal', MYR: 'Malaysian Ringgit', RON: 'Romanian Leu',
    DOP: 'Dominican Peso', BGN: 'Bulgarian Lev', HKD: 'Hong Kong Dollar',
    HRK: 'Croatian Kuna', PKR: 'Pakistani Rupee', EGP: 'Egyptian Pound',
    ISK: 'Icelandic Króna', VND: 'Vietnamese Dong', NGN: 'Nigerian Naira'
};

class CurrencyConverter {
    constructor() {
        this.cacheFile = CACHE_FILE;
        this.historyFile = HISTORY_FILE;
    }

    async loadCache() {
        try {
            const data = await fs.readFile(this.cacheFile, 'utf8');
            return JSON.parse(data);
        } catch (error) {
            return {};
        }
    }

    async saveCache(rates) {
        await fs.writeFile(this.cacheFile, JSON.stringify(rates, null, 2));
    }

    async saveToHistory(amount, fromCurrency, toCurrency, result) {
        try {
            let history = [];
            try {
                const data = await fs.readFile(this.historyFile, 'utf8');
                history = JSON.parse(data);
            } catch (error) {
                // File doesn't exist yet, start with empty history
            }

            history.push({
                timestamp: new Date().toISOString(),
                fromAmount: amount,
                fromCurrency,
                toCurrency,
                result
            });

            await fs.writeFile(this.historyFile, JSON.stringify(history, null, 2));
        } catch (error) {
            console.error(chalk.red('Error saving to history:', error.message));
        }
    }

    async showHistory() {
        try {
            const data = await fs.readFile(this.historyFile, 'utf8');
            const history = JSON.parse(data);

            const tableData = [
                ['Date', 'From', 'To', 'Result'].map(header => chalk.cyan(header))
            ];

            history.forEach(entry => {
                tableData.push([
                    new Date(entry.timestamp).toLocaleString(),
                    `${entry.fromAmount} ${entry.fromCurrency}`,
                    entry.toCurrency,
                    entry.result.toFixed(2)
                ]);
            });

            const config = {
                border: {
                    topBody: '─',
                    topJoin: '┬',
                    topLeft: '┌',
                    topRight: '┐',
                    bottomBody: '─',
                    bottomJoin: '┴',
                    bottomLeft: '└',
                    bottomRight: '┘',
                    bodyLeft: '│',
                    bodyRight: '│',
                    bodyJoin: '│',
                    joinBody: '─',
                    joinLeft: '├',
                    joinRight: '┤',
                    joinJoin: '┼'
                }
            };

            console.log(chalk.blue('\nConversion History'));
            console.log(table(tableData, config));
        } catch (error) {
            console.log(chalk.yellow('No conversion history found.'));
        }
    }

    listCurrencies() {
        const tableData = [
            ['Code', 'Currency Name'].map(header => chalk.cyan(header))
        ];

        Object.entries(CURRENCY_CODES)
            .sort(([a], [b]) => a.localeCompare(b))
            .forEach(([code, name]) => {
                tableData.push([chalk.green(code), name]);
            });

        const config = {
            border: {
                topBody: '─',
                topJoin: '┬',
                topLeft: '┌',
                topRight: '┐',
                bottomBody: '─',
                bottomJoin: '┴',
                bottomLeft: '└',
                bottomRight: '┘',
                bodyLeft: '│',
                bodyRight: '│',
                bodyJoin: '│',
                joinBody: '─',
                joinLeft: '├',
                joinRight: '┤',
                joinJoin: '┼'
            }
        };

        console.log(chalk.blue('\nAvailable Currencies'));
        console.log(table(tableData, config));
    }

    async convert(amount, fromCurrency, toCurrency, offline = false) {
        const spinner = ora('Converting currency...').start();

        try {
            let rate;
            if (offline) {
                const cache = await this.loadCache();
                const cacheKey = `${fromCurrency}_${toCurrency}`;
                rate = cache[cacheKey];

                if (!rate) {
                    spinner.fail();
                    console.error(chalk.red('No cached rate available. Please run in online mode first.'));
                    process.exit(1);
                }
            } else {
                const response = await axios.get(`${API_BASE_URL}/${fromCurrency}`);
                const rates = response.data.rates;
                rate = rates[toCurrency];

                if (!rate) {
                    spinner.fail();
                    console.error(chalk.red(`Rate not available for ${fromCurrency} to ${toCurrency}`));
                    process.exit(1);
                }

                // Cache the rate
                const cache = await this.loadCache();
                cache[`${fromCurrency}_${toCurrency}`] = rate;
                await this.saveCache(cache);
            }

            const result = amount * rate;
            spinner.succeed();
            return result;
        } catch (error) {
            spinner.fail();
            console.error(chalk.red('Error:', error.message));
            process.exit(1);
        }
    }
}

function displayResult(amount, fromCurrency, toCurrency, result) {
    const data = [
        [
            chalk.green(amount.toFixed(2)),
            chalk.yellow(fromCurrency),
            '→',
            chalk.green(result.toFixed(2)),
            chalk.yellow(toCurrency)
        ]
    ];

    const config = {
        border: {
            topBody: '─',
            topJoin: '┬',
            topLeft: '┌',
            topRight: '┐',
            bottomBody: '─',
            bottomJoin: '┴',
            bottomLeft: '└',
            bottomRight: '┘',
            bodyLeft: '│',
            bodyRight: '│',
            bodyJoin: '│',
            joinBody: '─',
            joinLeft: '├',
            joinRight: '┤',
            joinJoin: '┼'
        }
    };

    console.log(chalk.blue('\nCurrency Conversion Result'));
    console.log(table(data, config));
}

program
    .version('1.0.0')
    .description('Modern Currency Converter CLI Tool');

program
    .command('convert')
    .description('Convert currency from one to another')
    .requiredOption('--amount <number>', 'Amount to convert')
    .requiredOption('--from <currency>', 'Source currency code')
    .requiredOption('--to <currency>', 'Target currency code')
    .option('--save', 'Save conversion to history')
    .option('--offline', 'Use cached exchange rates')
    .action(async (options) => {
        const converter = new CurrencyConverter();
        const amount = parseFloat(options.amount);
        const fromCurrency = options.from.toUpperCase();
        const toCurrency = options.to.toUpperCase();

        // Validate currencies
        if (!CURRENCY_CODES[fromCurrency]) {
            console.error(chalk.red(`Error: Currency ${fromCurrency} is not supported`));
            process.exit(1);
        }
        if (!CURRENCY_CODES[toCurrency]) {
            console.error(chalk.red(`Error: Currency ${toCurrency} is not supported`));
            process.exit(1);
        }

        try {
            const result = await converter.convert(amount, fromCurrency, toCurrency, options.offline);
            displayResult(amount, fromCurrency, toCurrency, result);

            if (options.save) {
                await converter.saveToHistory(amount, fromCurrency, toCurrency, result);
                console.log(chalk.green('\nConversion saved to history.'));
            }
        } catch (error) {
            console.error(chalk.red('Error:', error.message));
            process.exit(1);
        }
    });

program
    .command('list')
    .description('List all available currencies')
    .action(() => {
        const converter = new CurrencyConverter();
        converter.listCurrencies();
    });

program
    .command('history')
    .description('Show conversion history')
    .action(async () => {
        const converter = new CurrencyConverter();
        await converter.showHistory();
    });

program.parse(process.argv); 